File: TreeBeautifulSet.java
import java.util.*;
import java.io.*;

/**
 * Tree Beautiful Set Problem Solution
 * 
 * This class implements a solution for finding beautiful sets in tree structures.
 * A beautiful set is a set of nodes where no node is an ancestor of another.
 * 
 * Key features:
 * - Tree construction from parent array
 * - DFS traversal for finding maximum beautiful sets
 * - Query processing for subtrees
 * - Modulo arithmetic (10^9 + 7)
 * 
 * @author Generated for Tree Beautiful Set Problem
 */
public class TreeBeautifulSet {
    
    private static final int MOD = 1000000007;
    private List<List<Integer>> adjacencyList;
    private int[] colors;
    private int nodeCount;
    
    /**
     * Constructor to initialize the tree structure
     * @param n Number of nodes in the tree
     */
    public TreeBeautifulSet(int n) {
        this.nodeCount = n;
        this.adjacencyList = new ArrayList<>();
        this.colors = new int[n];
        
        for (int i = 0; i < n; i++) {
            adjacencyList.add(new ArrayList<>());
        }
    }
    
    /**
     * Builds the tree from parent array
     * @param parents Array where parents[i] is the parent of node i (-1 for root)
     */
    public void buildTree(int[] parents) {
        for (int i = 0; i < parents.length; i++) {
            if (parents[i] != -1) {
                adjacencyList.get(parents[i]).add(i);
                adjacencyList.get(i).add(parents[i]);
            }
        }
    }
    
    /**
     * Sets colors for nodes
     * @param nodeColors Array of colors for each node
     */
    public void setColors(int[] nodeColors) {
        System.arraycopy(nodeColors, 0, this.colors, 0, nodeColors.length);
    }
    
    /**
     * DFS to find the maximum beautiful set in a subtree
     * A beautiful set contains nodes where no node is an ancestor of another
     * 
     * @param node Current node
     * @param parent Parent node (-1 if root)
     * @param memo Memoization map for dynamic programming
     * @return Pair of (include current node, exclude current node) maximum set sizes
     */
    public long[] dfsBeautifulSet(int node, int parent, Map<String, long[]> memo) {
        String key = node + "," + parent;
        if (memo.containsKey(key)) {
            return memo.get(key);
        }
        
        long includeNode = 1; // Include current node
        long excludeNode = 0; // Exclude current node
        
        // Process all children
        for (int child : adjacencyList.get(node)) {
            if (child != parent) {
                long[] childResult = dfsBeautifulSet(child, node, memo);
                
                // If we include current node, we can only exclude children
                includeNode = (includeNode + childResult[1]) % MOD;
                
                // If we exclude current node, we can take max of including or excluding children
                excludeNode = (excludeNode + Math.max(childResult[0], childResult[1])) % MOD;
            }
        }
        
        long[] result = {includeNode, excludeNode};
        memo.put(key, result);
        return result;
    }
    
    /**
     * Processes a query for finding beautiful sets in a subtree
     * @param rootNode Root of the subtree to process
     * @return Maximum size of beautiful set in the subtree
     */
    public long processQuery(int rootNode) {
        Map<String, long[]> memo = new HashMap<>();
        long[] result = dfsBeautifulSet(rootNode, -1, memo);
        return Math.max(result[0], result[1]) % MOD;
    }
    
    /**
     * Processes multiple queries
     * @param queries Array of root nodes for queries
     * @return Array of results for each query
     */
    public long[] processQueries(int[] queries) {
        long[] results = new long[queries.length];
        for (int i = 0; i < queries.length; i++) {
            results[i] = processQuery(queries[i]);
        }
        return results;
    }
    
    /**
     * Calculates sum of all query results modulo 10^9+7
     * @param queries Array of root nodes for queries
     * @return Sum of all query results modulo 10^9+7
     */
    public long calculateSumMod(int[] queries) {
        long sum = 0;
        for (int query : queries) {
            sum = (sum + processQuery(query)) % MOD;
        }
        return sum;
    }
    
    /**
     * Main method for testing the implementation
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        try {
            // Read number of nodes
            System.out.print("Enter number of nodes: ");
            int n = scanner.nextInt();
            
            TreeBeautifulSet tbs = new TreeBeautifulSet(n);
            
            // Read parent array
            System.out.println("Enter parent array (-1 for root): ");
            int[] parents = new int[n];
            for (int i = 0; i < n; i++) {
                parents[i] = scanner.nextInt();
            }
            tbs.buildTree(parents);
            
            // Read colors (optional)
            System.out.println("Enter colors for nodes: ");
            int[] colors = new int[n];
            for (int i = 0; i < n; i++) {
                colors[i] = scanner.nextInt();
            }
            tbs.setColors(colors);
            
            // Read number of queries
            System.out.print("Enter number of queries: ");
            int q = scanner.nextInt();
            
            int[] queries = new int[q];
            System.out.println("Enter query root nodes: ");
            for (int i = 0; i < q; i++) {
                queries[i] = scanner.nextInt();
            }
            
            // Process queries and calculate sum
            long result = tbs.calculateSumMod(queries);
            System.out.println("Sum of all beautiful set sizes (mod 10^9+7): " + result);
            
            // Individual query results
            long[] individualResults = tbs.processQueries(queries);
            System.out.println("Individual query results:");
            for (int i = 0; i < q; i++) {
                System.out.println("Query " + (i+1) + " (root " + queries[i] + "): " + individualResults[i]);
            }
            
        } catch (Exception e) {
            System.err.println("Error processing input: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }
    
    /**
     * Utility method to print tree structure for debugging
     */
    public void printTree() {
        System.out.println("Tree structure:");
        for (int i = 0; i < nodeCount; i++) {
            System.out.print("Node " + i + " -> ");
            for (int neighbor : adjacencyList.get(i)) {
                System.out.print(neighbor + " ");
            }
            System.out.println();
        }
    }
    
    /**
     * Test method with sample data
     */
    public static void runTests() {
        System.out.println("Running test cases...");
        
        // Test Case 1: Simple tree
        TreeBeautifulSet test1 = new TreeBeautifulSet(5);
        int[] parents1 = {-1, 0, 0, 1, 1}; // Tree: 0 is root, 1,2 are children of 0, 3,4 are children of 1
        test1.buildTree(parents1);
        test1.setColors(new int[]{1, 2, 1, 3, 2});
        
        System.out.println("Test 1 - Tree structure:");
        test1.printTree();
        
        int[] queries1 = {0, 1, 2};
        long[] results1 = test1.processQueries(queries1);
        System.out.println("Test 1 Results:");
        for (int i = 0; i < queries1.length; i++) {
            System.out.println("Query from root " + queries1[i] + ": " + results1[i]);
        }
        System.out.println("Sum: " + test1.calculateSumMod(queries1));
        
        System.out.println();
        
        // Test Case 2: Linear tree
        TreeBeautifulSet test2 = new TreeBeautifulSet(4);
        int[] parents2 = {-1, 0, 1, 2}; // Linear: 0->1->2->3
        test2.buildTree(parents2);
        test2.setColors(new int[]{1, 1, 1, 1});
        
        System.out.println("Test 2 - Linear tree:");
        test2.printTree();
        
        int[] queries2 = {0, 1};
        long[] results2 = test2.processQueries(queries2);
        System.out.println("Test 2 Results:");
        for (int i = 0; i < queries2.length; i++) {
            System.out.println("Query from root " + queries2[i] + ": " + results2[i]);
        }
        System.out.println("Sum: " + test2.calculateSumMod(queries2));
    }
}

File: tree_beautiful_set.py
"""
Tree Beautiful Set Problem Solution

This module implements a solution for finding beautiful sets in tree structures.
A beautiful set is defined as a set of nodes where no two nodes are adjacent
and all nodes have the same color.

Problem Description:
- Given a tree with N nodes, parent array, colors array, and Q queries
- Each query asks for the sum of maximum beautiful set sizes in subtrees
- Return the sum modulo 10^9 + 7

Author: Generated for internship1 repository
"""

from collections import defaultdict
from typing import List, Dict, Tuple, Optional


class TreeBeautifulSet:
    """
    A class to handle tree operations and queries for finding beautiful sets.
    
    A beautiful set is a set of nodes where:
    1. No two nodes in the set are adjacent (connected by an edge)
    2. All nodes in the set have the same color
    """
    
    MOD = 10**9 + 7
    
    def __init__(self):
        """Initialize the TreeBeautifulSet solver."""
        self.n = 0
        self.adj = defaultdict(list)
        self.colors = []
        self.parent = []  # Store parent information
        
    def build_tree(self, n: int, parents: List[int], colors: List[int]) -> None:
        """
        Build the adjacency list representation of the tree.
        
        Args:
            n (int): Number of nodes in the tree
            parents (List[int]): Parent array where parents[i] is parent of node i
                                (parents[0] should be -1 for root)
            colors (List[int]): Color array where colors[i] is color of node i
        """
        self.n = n
        self.colors = colors[:]
        self.parent = parents[:]
        self.adj = defaultdict(list)
        
        # Build adjacency list from parent array
        for i in range(n):
            if parents[i] != -1:
                self.adj[parents[i]].append(i)
                self.adj[i].append(parents[i])
    
    def find_max_beautiful_set(self, root: int) -> int:
        """
        Find the maximum beautiful set size in the subtree rooted at the given node.
        
        Args:
            root (int): Root node of the subtree
            
        Returns:
            int: Maximum beautiful set size in the subtree
        """
        if root >= self.n or root < 0:
            return 0
        
        # Get all nodes in the subtree rooted at 'root'
        # Use the parent information to determine the correct parent
        root_parent = self.parent[root]
        
        subtree_nodes = []
        
        def collect_subtree_nodes(node, parent):
            subtree_nodes.append(node)
            for child in self.adj[node]:
                if child != parent:
                    collect_subtree_nodes(child, node)
        
        collect_subtree_nodes(root, root_parent)
        
        # Get unique colors in this subtree only
        unique_colors = set(self.colors[node] for node in subtree_nodes)
        
        max_size = 0
        
        # For each color, find the maximum beautiful set of that color in this subtree
        for color in unique_colors:
            color_max = self._solve_for_color(root, root_parent, color, set(subtree_nodes))
            max_size = max(max_size, color_max)
        
        return max_size
    
    def _solve_for_color(self, node: int, parent: int, target_color: int, subtree_nodes: set) -> int:
        """
        Solve the maximum independent set problem for nodes of a specific color.
        
        This is a classic tree DP problem: for each node, we decide whether to include it or not.
        
        Args:
            node (int): Current node
            parent (int): Parent node (-1 if root)
            target_color (int): Target color for the beautiful set
            subtree_nodes (set): Set of all nodes in the current subtree
            
        Returns:
            int: Maximum beautiful set size for this color starting from this node
        """
        # If node is not in current subtree, return 0
        if node not in subtree_nodes:
            return 0
            
        # Get children in the subtree
        children = [child for child in self.adj[node] if child != parent and child in subtree_nodes]
        
        # Base case: leaf node
        if not children:
            return 1 if self.colors[node] == target_color else 0
        
        # Option 1: Include current node (only if it has target color)
        include_current = 0
        if self.colors[node] == target_color:
            include_current = 1
            # If we include current node, we cannot include any children
            # But we can include grandchildren
            for child in children:
                grandchildren = [gc for gc in self.adj[child] if gc != node and gc in subtree_nodes]
                for grandchild in grandchildren:
                    include_current += self._solve_for_color(grandchild, child, target_color, subtree_nodes)
        
        # Option 2: Exclude current node
        exclude_current = 0
        for child in children:
            exclude_current += self._solve_for_color(child, node, target_color, subtree_nodes)
        
        return max(include_current, exclude_current)
    
    def process_queries(self, queries: List[int]) -> int:
        """
        Process multiple queries and return the sum of results.
        
        Args:
            queries (List[int]): List of root nodes for subtree queries
            
        Returns:
            int: Sum of maximum beautiful set sizes for all queries modulo MOD
        """
        total_sum = 0
        
        for root in queries:
            max_beautiful = self.find_max_beautiful_set(root)
            total_sum = (total_sum + max_beautiful) % self.MOD
        
        return total_sum
    
    def solve(self, n: int, parents: List[int], colors: List[int], queries: List[int]) -> int:
        """
        Complete solution method that builds the tree and processes queries.
        
        Args:
            n (int): Number of nodes
            parents (List[int]): Parent array
            colors (List[int]): Colors array
            queries (List[int]): Query nodes
            
        Returns:
            int: Sum of results modulo MOD
        """
        self.build_tree(n, parents, colors)
        return self.process_queries(queries)


def solve_tree_beautiful_set(n: int, parents: List[int], colors: List[int], queries: List[int]) -> int:
    """
    Convenience function to solve the Tree Beautiful Set problem.
    
    Args:
        n (int): Number of nodes in the tree
        parents (List[int]): Parent array where parents[i] is parent of node i
        colors (List[int]): Color array where colors[i] is color of node i
        queries (List[int]): List of root nodes for subtree queries
        
    Returns:
        int: Sum of maximum beautiful set sizes for all queries modulo 10^9+7
    
    Example:
        >>> n = 5
        >>> parents = [-1, 0, 0, 1, 1]  # Tree: 0 -> [1, 2], 1 -> [3, 4]
        >>> colors = [1, 1, 2, 1, 2]   # Colors for each node
        >>> queries = [0, 1]           # Query subtrees rooted at 0 and 1
        >>> result = solve_tree_beautiful_set(n, parents, colors, queries)
        >>> print(result)
    """
    solver = TreeBeautifulSet()
    return solver.solve(n, parents, colors, queries)


# Input/Output handling functions
def read_input() -> Tuple[int, List[int], List[int], List[int]]:
    """
    Read input from standard input.
    
    Expected format:
    Line 1: n (number of nodes)
    Line 2: n space-separated integers (parent array)
    Line 3: n space-separated integers (colors array)
    Line 4: q (number of queries)
    Line 5: q space-separated integers (query nodes)
    
    Returns:
        Tuple containing n, parents, colors, and queries
    """
    n = int(input().strip())
    parents = list(map(int, input().strip().split()))
    colors = list(map(int, input().strip().split()))
    q = int(input().strip())
    queries = list(map(int, input().strip().split()))
    
    return n, parents, colors, queries


def main():
    """Main function for command-line usage."""
    try:
        n, parents, colors, queries = read_input()
        result = solve_tree_beautiful_set(n, parents, colors, queries)
        print(result)
    except Exception as e:
        print(f"Error: {e}")


if __name__ == "__main__":
    main()